src/Microsoft.Identity.Web.DownstreamRestApi/DownstreamRestApi.cs

```diff
@@ -326,44 +216,68 @@ internal class DownstreamRestApi : IDownstreamRestApi
+            string? serviceName,
+            DownstreamRestApiOptions effectiveOptions,
+            bool appToken,
+            HttpContent? content = null,
+            ClaimsPrincipal? user = null,
+            CancellationToken cancellationToken = default)
         {
-            Action<DownstreamRestApiOptions> downstreamRestApiOptions;
-
-            if (downstreamRestApiOptionsOverride == null)
+            if (effectiveOptions.Scopes == null)
             {
-                downstreamRestApiOptions = options =>
-                {
-                    options.HttpMethod = httpMethod;
-                };
+                throw new ArgumentException(ScopesNotConfiguredInConfigurationOrViaDelegate);
             }
-            else
+
+            // Downstream API URI
+            string apiUrl = effectiveOptions.GetApiUrl();
+
+            // Creation of the Http Request message with the right HTTP Method
+            using HttpRequestMessage httpRequestMessage = new HttpRequestMessage(
+                effectiveOptions.HttpMethod,
+                apiUrl);
+            if (content != null)
             {
-                downstreamRestApiOptions = options =>
-                {
-                    downstreamRestApiOptionsOverride(options);
-                    options.HttpMethod = httpMethod;
-                };
+                httpRequestMessage.Content = content;
             }
 
-            return downstreamRestApiOptions;
-        }
+            // Obtention of the authorization header
+            string authorizationHeader = appToken ?
+                await _authorizationHeaderProvider.CreateAuthorizationHeaderForAppAsync(
+                                        effectiveOptions.Scopes.FirstOrDefault(),
+                                        effectiveOptions,
+                                        cancellationToken).ConfigureAwait(false) :
+                await _authorizationHeaderProvider.CreateAuthorizationHeaderForUserAsync(
+                                        effectiveOptions.Scopes,
+                                        effectiveOptions,
+                                        user,
+                                        cancellationToken).ConfigureAwait(false);
+            httpRequestMessage.Headers.Add(Authorization, authorizationHeader);
 
-        private static StringContent ConvertFromInput<TInput>(TInput input)
-        {
-            return new StringContent(JsonSerializer.Serialize(input), Encoding.UTF8, "application/json");
+            // Opportunity to change the request message
+            effectiveOptions.CustomizeHttpRequestMessage?.Invoke(httpRequestMessage);
+
+            // Send the HTTP message
+            using HttpClient client = string.IsNullOrEmpty(serviceName) ? _httpClientFactory.CreateClient() : _httpClientFactory.CreateClient(serviceName);
+            return await client.SendAsync(httpRequestMessage, cancellationToken).ConfigureAwait(false);
```

Comment:
do we need to lock on the HttpClient?

Reply:
no I don't think so. Usually we lock on the HttpContext. Do we lock on the HttpClient?s